### PR TITLE
Migrate from HPC cloud VMs to Research Cloud application

### PR DESCRIPTION
**Pull request by sverhoeven**
The plugin, application and application offer where made in Research Cloud which should run the `research-cloud-plugin.yml` playbook.

A second playbook called `shared-data-disk.yml` has been made that can download the data from Internet and upload it to dcache.

Refs #77

### CODE REVIEW

**Comment by sverhoeven** [comment]
Application can be instantiated and logged in with ssh. However visiting home page fails with bad gateway.

**Comment by sverhoeven** [comment]
Successfully ran https://github.com/eWaterCycle/ewatercycle/blob/main/docs/examples/MarrmotM01.ipynb notebook only required change was to set  directory="/data/volume_2/forcings/MARRMoT".

**Review by Peter9192 — COMMENTED** [review]
Some initial comments, will continue reviewing today

File: roles/launcher/tasks/main.yml [hunk 1]
```diff
@@ -40,3 +41,23 @@
...
+      location /launcher {
+          proxy_pass http://127.0.0.1:8888;
+          proxy_redirect http://localhost:8888/ $scheme://$host/;
+          proxy_set_header X-Real-IP $remote_addr;
+          proxy_set_header Host $host;
+          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+          proxy_set_header Upgrade $http_upgrade;
+          proxy_set_header Connection $connection_upgrade;
+          client_max_body_size 10G;
+      }
```
Comment (Peter9192) [inline roles/launcher/tasks/main.yml:63]: Similar to /roles/jupyter/tasks/hub? This redirects the user to their own notebook server or so?

File: shared-data-disk.yml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        - docker: ewatercycle/pcrg-grpc4bmi:setters
+          singularity: ewatercycle-pcrg-grpc4bmi_setters.sif
+        - docker: ewatercycle/wflow-grpc4bmi:2020.1.1
+          singularity: ewatercycle-wflow-grpc4bmi_2020.1.1.sif
+        - docker:  ewatercycle/marrmot-grpc4bmi:2020.11
+          singularity: ewatercycle-marrmot-grpc4bmi_2020.11.sif
+        # - docker: ewatercycle/hype-grpc4bmi
+        #   singularity: ewatercycle-hype-grpc4bmi.sif
+        - docker: ewatercycle/lisflood-grpc4bmi:20.10
+          singularity: ewatercycle-lisflood-grpc4bmi_20.10.sif
```
Comment (Peter9192) [inline shared-data-disk.yml:26]: Duplicate of  infra/roles/jupyter/defaults/main.yml ?

File: src-questions.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Q for SRC support
+
+I am writing my own plugin where I need to set the public URL in an application config file using an Ansible task.
+In the logs of a workspace I see a `service_url` key with value `https://d08d27c8-d63d-492f-9479-70aa98b90f81.workspaces.live.surfresearchcloud.nl`.
```
Comment (Peter9192) [inline src-questions.md:4]: Should we gather all todo's, issues, and questions here?

File: roles/launcher/templates/experiment-launcher.systemd.j2 [hunk 4]
```diff
@@ -6,7 +6,7 @@ After=syslog.target network.target
 
 [Service]
 User=root
-Environment=JUPYTERHUB_TOKEN={{ launcher_jupyterhub_token }}
+Environment=JUPYTERHUB_TOKEN={{ lookup('password', 'jupyterhub.launcher.token length=32 chars=ascii_letters,digits') }}
```
Comment (Peter9192) [inline roles/launcher/templates/experiment-launcher.systemd.j2:9]: How does that lookup work?

File: roles/jupyter/templates/jupyterhub_config.py.j2 [hunk 5]
```diff
@@ -7,7 +7,7 @@ c.JupyterHub.services = [
     {
         'name': 'experiment-launcher',
         'admin': True,
-        'api_token': '{{ launcher_jupyterhub_token }}',
+        'api_token': '{{ lookup('password', 'jupyterhub.launcher.token length=32 chars=ascii_letters,digits') }}',
```
Comment (Peter9192) [inline roles/jupyter/templates/jupyterhub_config.py.j2:10]: How does this work?

**Review by Peter9192 — COMMENTED** [review]
Congrats @sverhoeven! That's a lot of work you did there. It seems to work great for what I've tested. The only thing I'm not entirely confident about is that rclone caching mechanism we talked about. 

I left a couple of comments/questions. Perhaps we can go over the PR together so you don't have to type out all the answers. 

Notes from testing:

- `model.initialize` timed out a few times, probably should increase the timeout limit
- explorer 'start experiment' doesn't work (but has been updated since?)
- `ls /mnt/data/esmvaltool/obs6/Tier3/ERA5` lists '=3'
- not sure if `/mnt/data/esmvaltool/` is a good name. Perhaps `/climatedata` or `/forcing_data`?

Further comments on the PR

- A bit more elaborate readme's might help to understand what each role is responsible for
- Is this also a natural moment to clean up the attic?

File: research-cloud-plugin.yml [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        dest: /etc/rclone.conf
+        mode: 0644
+        content: |
+          [{{ dcache_rclone_name }}]
+          type = webdav
+          url = {{ dcache_url }}
+          vendor = other
+          user =
+          pass =
+          bearer_token = {{ dcache_ro_token }}
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply research-cloud-plugin.yml:120]: Done in latest commit

File: research-cloud-plugin.yml [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+      debug:
+        msg: The eWaterCycle Jupyter plugin has completed
+
+# TODO Stuff to ask Research Cloud helpdesk
+# 2. How to become root after creation, see https://servicedesk.surfsara.nl/wiki/display/WIKI/RC+plugin+passwordless+sudo
+# 3. How to add users to docker posix group
+# 6. How to mark certain user as JupyterHub admin -> use sram api to sync posix group
+# 9. How to put home dirs on attached storage
+# 11. What is symlink to /scratch doing in homedir? How to tie storage item to it?
+# 12. Fetch workspace.json from https//gw.live.surfresearchcloud.nl/workspace/workspaces gives "Access to this API has been disallowed" error
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply research-cloud-plugin.yml]: Moved to src-questions.md

**Review by Peter9192 — APPROVED** [review]
Looks very good! Thanks for the added documentation. Let's merge this and create follow-up PRs for remaining tasks.

Can we move the remaining todo's to issues?

**Comment by sverhoeven** [comment]
Thanks for reviewing.

**Pull request merged**